examples/Fine_tuning_for_function_calling.ipynb

```diff
@@ -0,0 +1,1274 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Generating synthetic data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Helper functions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We want to generate every invocation of every function, so that we have\n",
+    "full coverage of all potential invocations to create synthetic data for. Then, we will use GPT-4 to come up with prompts that would call each invocation, and we will use that prompt - function invocation pair as training data."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Generating every invocation for a function with fixed enums is more simple, but for a function such as\n",
+    "  `control_gimbal` we need to set the `tilt` and `pan` values which are integers, so to generate those synthetic invocations we will first set a place holder, and then later use GPT-4 to come up with reasonable values."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "placeholder_int = 'fill_in_int'\n",
+    "placeholder_string = 'fill_in_string'\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The functions below take in all the functions from the function list, and look \n",
+    "at all the potential invocations of those functions given each function's parameters.\n",
+    "The functions also account for `required` parameters, so that all the invocations\n",
+    "are actually feasible."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def generate_permutations(params: Dict[str, Dict[str, Any]]) -> Generator[Dict[str, Any], None, None]:\n",
+    "  required_fields = params.get('required', [])\n",
+    "  required_permutations = generate_required_permutations(params, required_fields)\n",
+    "  for required_perm in required_permutations:\n",
+    "      yield from generate_optional_permutations(params, required_perm)\n",
+    "\n",
+    "def generate_required_permutations(params: Dict[str, Dict[str, Any]], required_fields: List[str]) -> List[Dict[str, Any]]:\n",
+    "  required_values = [get_possible_values(params, field) for field in required_fields]\n",
+    "  return [dict(zip(required_fields, values)) for values in itertools.product(*required_values)]\n",
+    "\n",
+    "def generate_optional_permutations(params: Dict[str, Dict[str, Any]], base_perm: Dict[str, Any]) -> Generator[Dict[str, Any], None, None]:\n",
+    "  optional_fields = set(params['properties']) - set(base_perm)\n",
+    "  for field_subset in itertools.chain.from_iterable(itertools.combinations(optional_fields, r) for r in range(len(optional_fields)+1)):\n",
```

Comment:
This code is a bit hard to parse, let's maybe add some comments and some spacing. An autoformatter might help. You can also split things into functions. I bet GPT-4 has some ideas for how to make it more readable.